src/prpy/planning/base.py

```diff
@@ -332,62 +309,44 @@ def get_planners(self, method_name):
 
     def plan(self, method, args, kw_args):
         all_planners = self._planners
-        planners = []
+        futures = dict()
         results = [None] * len(self._planners)
 
+        # Helper function to call a planner and return its result.
+        def call_planner(planner):
+            planning_method = getattr(planner, method)
+            return planning_method(*args, **kw_args)
+
         # Find only planners that support the required planning method.
+        # Call every planners in parallel using a concurrent executor and
+        # return the first non-error result in the ordering when available.
         for index, planner in enumerate(all_planners):
             if not planner.has_planning_method(method):
                 results[index] = PlanningError(
                     "{:s} does not implement method {:s}."
-                    .format(planner, method)
-                )
+                    .format(planner, method))
                 continue
             else:
-                planners.append((index, planner))
+                futures[index] = defer(call_planner, args=(planner,))
 
-        # Helper function to call a planner and store its result or error.
-        def call_planner(index, planner):
+        # Each time a planner completes, check if we have a valid result
+        # (a planner found a solution and all higher-ranked planners had
+        # already failed).
+        for index, future in futures.viewitems():
```

Comment:
Python `dict`s are not ordered by key, so this *does not* call `result()` on the futures in the correct order. I suggest replacing `futures` with a `list` of `tuples` since you never query by index.